add http doc (#56)
minetest/minetest_docs: doc/http_api.md

# HTTP API

Enables mods to make HTTP requests.

## Settings

> To use this api, mods need to be listed in one or the other following 

Reviewer: ```suggestion
To use this API, mods need to be listed in one or the other following 
```

Reviewer: Applies throughout the file

> settings in csv format:

Reviewer: ```suggestion
settings in CSV format:
```

* `secure.trusted_mods`
* `secure.http_mods`

An example would be:

```
secure.http_mods = modname1, modnametwo, modenamethree
```

## Types

### HttpRequest

A table consisting of the following paramaters:

> * url (required): 

Reviewer: I would format any code value in backticks for clarity

```suggestion
* `url` (required): 
```

Author: https://github.com/minetest/minetest_docs/pull/56/commits/31405bcab27031ae5769c38af5f44d13a5beb7e7

>   * `http://subdomain.domain.tld`

Reviewer: Should we use https by default in examples? Is any special code needed to connect to http vs https?

```suggestion
  * `https://subdomain.domain.tld`
```

>   * `https://subdomain.domain.tld`

Author: https://github.com/minetest/minetest_docs/pull/56/commits/e7209e47bfd967f11b79bb5b0f6ae7a6c507c96e

* `timeout` (optional): 
  * Time in seconds
  * Default is what `curl_timeout` setting is set to (NOTE: setting is 
    in milliseconds)
* `method` (optional):

>   * Supports `GET`, `POST`, `PUT`, `DELETE`

Author: https://github.com/minetest/minetest_docs/pull/56/commits/e7209e47bfd967f11b79bb5b0f6ae7a6c507c96e

  * Default is `GET`
* `data` (optional):
  * Supports string or table
  * Tables are encoded as `x-www-form-urlencoded` key value pairs
* `post_data` (optional):
  * Deprecated, see data above
* `user_agent` (optional):
  * Replaces the default user agent if provided
* `extra_headers` (optional):

>   * Table of strings

Author: its a type of table

  * Each string needs to follow http spec, `"key: value"`
* multipart (optional):
  * Boolean, default false
  * Method must be `POST`

>   * [see here for more info](https://en.wikipedia.org/wiki/MIME#Multipart_messages)

Reviewer: Text of a link should be a bit unique to the link for readability, some screen reader users will use keyboard shortcuts to read the text of each link. If all links have similar text, it's hard to use.

```suggestion
  * Learn more at [MIME (Wikpedia)](https://en.wikipedia.org/wiki/MIME#Multipart_messages)
```

### HttpRequestResult

> A table consisting of the following paramaters

Reviewer 2: ```suggestion
A table consisting of the following parameters
```

* `completed`:
  * Boolean

>   * If requested finished

Reviewer 2: What precisely does "finished" mean here? The engine docs are clearer about this:

> (either succeeded, failed or timed out)

So basically the purpose of this field is to enable polling.

Author: https://github.com/minetest/minetest_docs/pull/56/commits/d8315ffea2abef8748e090468082a81511e0369e

* `succeeded`:
  * Boolean
  * Request status
* `timeout`:
  * If it timed out or not, see HttpRequest timeout
* `code`:
  * Int
  * Http status code
* `data`:
  * Table or string
  * Response returned

## Functions

### core.request_http_api()

* Returns a table containing the following functions provided:

>   * That the mod in question has been placed in on of the two 

Reviewer 2: ```suggestion
  * That the mod in question has been listed in one of the two 
```

    settings above
  * The engine was built with curl support

>   * Called from the mods init.lua file and not in a function

Reviewer 2: ```suggestion
  * `core.request_http_api()` is called from the mod's `init.lua` file and not in a function
```

Author: was going to say something, but i suppose it cant hurt to be more explicit

> [!WARNING]  

> > Store the result in a local variable, not a global so that 

Reviewer: ```suggestion
> Store the result in a local variable, not a global, so that 
```

> > other mods cannot access it, they should request their own 

Reviewer: ```suggestion
> other mods cannot access it. Other mods should request their own 
```

> > access

Reviewer 2: ```suggestion
> access.
```

> An example of how to use this api (handling both existence and permission 

Reviewer: ```suggestion
An example of how to use this API (handling both existence and permission 
```

being granted) is as follows:
```lua
    --inside mods init.lua file

>     local http = core.request_http_api and core.request_http_api()

Reviewer 2: Why is this indented?

Reviewer 2: I would leave such style decisions up to the stylesheets this is ultimately rendered with.

(What I also don't like about the indentation is that if your `core.request_http_api` was in a function, it'd be indented as well. So I fear this might nudge readers in the wrong direction.)

    if http then
        --call your http methods here
    end
```

> ### httpapi.fetch(request, callback)

Reviewer: Is the case-difference of `httpapi` and `HttpApi` intentional and necessary? If so, let's document it, otherwise let's use a consistent case

```suggestion
### HttpApi.fetch(request, callback)
```

> Calls callback once completed

Reviewer 2: Maybe state that using this method is recommended?

Author: https://github.com/minetest/minetest_docs/pull/56/commits/43a9c4ff8b0bb7260b6d8379cd2694d1e0c5cbb6

* Request is of type HttpRequest

> * Callback recieves one param of type HttpRequestResult

Reviewer 2: ```suggestion
* Callback receives one param of type `HttpRequestResult`
```

settings in CSV format:
* `secure.trusted_mods`
* `secure.http_mods`

An example would be:

```
secure.http_mods = modname1, modnametwo, modenamethree
```

## Types

### HttpRequest

A table consisting of the following paramaters:

* `url` (required):
  * Type string
  * Example: `https://subdomain.domain.tld`
* `timeout` (optional): 
  * Time in seconds
  * Default is what `curl_timeout` setting is set to (NOTE: setting is 
    in milliseconds)
* `method` (optional):
  * Type string
  * Supports `GET`, `POST`, `PUT`, `DELETE`
  * Default is `GET`
* `data` (optional):
  * Supports string or table
  * Tables are encoded as `x-www-form-urlencoded` key value pairs
* `post_data` (optional):
  * Deprecated, see data above
* `user_agent` (optional):
  * Type string
  * Replaces the default user agent if provided
* `extra_headers` (optional):
  * Table of strings
  * Each string needs to follow http spec, `"key: value"`
* `multipart` (optional):
  * Boolean, default false
  * Method must be `POST`
  * Learn more at [MIME (Wikpedia)](https://en.wikipedia.org/wiki/MIME#Multipart_messages)

### HttpRequestResult

A table consisting of the following parameters

* `completed`:
  * Boolean

>   * If requested finished. By way of success, failure, or timing out

Reviewer 2: ```suggestion
  * If request finished. By way of success, failure, or timing out.
```
